src/lib/poracleMessage/commands/egg.js

```diff
@@ -133,12 +132,14 @@ exports.run = async (client, msg, args, options) => {
 				})
 			}
 
-			if (insert.length) {
```

Comment:
This fix was because insertQuery fails with an empty array, and if all the additions are eliminated (because they go into the unchanged list), then it was causing an exception. I think the same condition exists in your revised code, but now check should be ... `if (insert.length || updates.length)

Reply:
why aren't we using upsert for tracks any more?

Reply:
Too many columns for mysql to support the required index

Reply:
was a depressing time, mysql and mariadb have different limits in this area too :-(

Reply:
I don't know why db methods are duplicated between controller.js and query.js but I've added the crutch in a single place instead of each command

Reply:
Short answer - start of some refactoring not yet complete.

Longer answer - controller.js being the base class for all of the alarm controllers was quite large and had many more dependencies than were really needed by simple database work.  So first part of refactoring was to move out the database access which is used more broadly (outside of alarms in many places).
It turns out that the commands use some other things which aren't queries which had to be moved across also.
But these should really be split out, and then controller.js pointed to query so that db access is through one place.

Crutch makes sense.